unit flake: pkg/client/unversioned/portforward is timing out

FAIL	k8s.io/kubernetes/pkg/client/unversioned/portforward	300.229s

Example failure:
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/e41d66f0e207f8d32268b80e8d0d1c227663cc48/kubernetes-pull-test-unit-integration/7816/artifacts/?debugUI=DEVELOPERS

Pasting the log from 17916:


```
=== RUN   TestGetListener
I0101 21:35:33.696843   12432 portforward.go:213] Forwarding from localhost:12345 -> 12345
I0101 21:35:33.697201   12432 portforward.go:213] Forwarding from 127.0.0.1:12345 -> 12345
I0101 21:35:33.697423   12432 portforward.go:213] Forwarding from [::1]:12345 -> 12345
E0101 21:35:33.697685   12432 portforward.go:202] Unable to create listener: Error listen tcp4 [::1]:12345: non-IPv4 address <nil>
E0101 21:35:33.697873   12432 portforward.go:202] Unable to create listener: Error listen tcp6 127.0.0.1:12345: bind: invalid argument
E0101 21:35:33.698001   12432 portforward.go:202] Unable to create listener: Error listen tcp6: too many colons in address ::1:12345
--- PASS: TestGetListener (0.00s)
    portforward_test.go:193: Asked a tcp4 forward for: localhost:12345, got listener 127.0.0.1:12345, expected: 12345
    portforward_test.go:193: Asked a tcp4 forward for: 127.0.0.1:12345, got listener 127.0.0.1:12345, expected: 12345
    portforward_test.go:193: Asked a tcp6 forward for: [::1]:12345, got listener ::1:12345, expected: 12345
=== RUN   TestForwardPorts
I0101 21:35:33.708779   12432 portforward.go:213] Forwarding from 127.0.0.1:5000 -> 5000
I0101 21:35:33.709130   12432 portforward.go:213] Forwarding from [::1]:5000 -> 5000
panic: test timed out after 5m0s

goroutine 12 [running]:
testing.startAlarm.func1()
    /tmp/workdir/go/src/testing/testing.go:703 +0x16b
created by time.goFunc
    /tmp/workdir/go/src/time/sleep.go:129 +0x6e

goroutine 1 [chan receive]:
testing.RunTests(0x17f40e0, 0x1e32640, 0x4, 0x4, 0xc8201def01)
    /tmp/workdir/go/src/testing/testing.go:562 +0xafa
testing.(*M).Run(0xc8200d9ef0, 0x4861bf)
    /tmp/workdir/go/src/testing/testing.go:494 +0xe5
main.main()
    k8s.io/kubernetes/pkg/client/unversioned/portforward/_test/_testmain.go:60 +0x210

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
    /tmp/workdir/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 19 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x1e537e0)
    /workspace/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x8c
created by github.com/golang/glog.init.1
    /workspace/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x2ba

goroutine 7 [semacquire]:
sync.runtime_Semacquire(0xc8202050bc)
    /tmp/workdir/go/src/runtime/sema.go:43 +0x26
sync.(*WaitGroup).Wait(0xc8202050b0)
    /tmp/workdir/go/src/sync/waitgroup.go:126 +0x118
net/http/httptest.(*Server).Close(0xc820205080)
    /tmp/workdir/go/src/net/http/httptest/server.go:168 +0x81
k8s.io/kubernetes/pkg/client/unversioned/portforward.TestForwardPorts(0xc8201e02d0)
    /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/unversioned/portforward/portforward_test.go:360 +0x1230
testing.tRunner(0xc8201e02d0, 0x1e32670)
    /tmp/workdir/go/src/testing/testing.go:456 +0xdd
created by testing.RunTests
    /tmp/workdir/go/src/testing/testing.go:561 +0xaa4

goroutine 24 [chan send]:
k8s.io/kubernetes/pkg/util/httpstream/spdy.(*connection).SetIdleTimeout(0xc82031a180, 0x0)
    /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/httpstream/spdy/connection.go:138 +0x9b
k8s.io/kubernetes/pkg/kubelet/server.ServePortForward(0x7f38e6bec0d0, 0xc8201b4580, 0xc8201aae00, 0x7f38e6bec108, 0xc8202012e0, 0x1598378, 0x3, 0x1599e40, 0x3, 0x0, ...)
    /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/server/server.go:796 +0x4b6
k8s.io/kubernetes/pkg/client/unversioned/portforward.fakePortForwardServer.func1(0x7f38e6bec0d0, 0xc8201b4580, 0xc8201aae00)
    /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/unversioned/portforward/portforward_test.go:253 +0x265
net/http.HandlerFunc.ServeHTTP(0xc8201dfe00, 0x7f38e6bec0d0, 0xc8201b4580, 0xc8201aae00)
    /tmp/workdir/go/src/net/http/server.go:1422 +0x48
net/http/httptest.(*waitGroupHandler).ServeHTTP(0xc820201120, 0x7f38e6bec0d0, 0xc8201b4580, 0xc8201aae00)
    /tmp/workdir/go/src/net/http/httptest/server.go:200 +0xff
net/http.serverHandler.ServeHTTP(0xc820011ec0, 0x7f38e6bec0d0, 0xc8201b4580, 0xc8201aae00)
    /tmp/workdir/go/src/net/http/server.go:1862 +0x207
net/http.(*conn).serve(0xc820316000)
    /tmp/workdir/go/src/net/http/server.go:1361 +0x117d
created by net/http.(*Server).Serve
    /tmp/workdir/go/src/net/http/server.go:1910 +0x465
FAIL    k8s.io/kubernetes/pkg/client/unversioned/portforward    300.274s
```

I believe what's happening here is the client is opening and closing the spdy connection so quickly at times that the spdy server is shutting down its goroutines just as quickly, before our code is able to call `SetIdleTimeout`. And when our code does call that, it hangs because there is nothing reading from the channel that `SetIdleTimeout` is sending to.

--- a/Godeps/Godeps.json
+++ b/Godeps/Godeps.json
@@ -394,7 +394,7 @@
 		},
 		{
 			"ImportPath": "github.com/docker/spdystream",
-			"Rev": "43bffc458d55aa784be658c9867fbefcfcb7fecf"
+			"Rev": "c33989bcb56748d2473194d11f8ac3fc563688eb"
 		},
 		{
 			"ImportPath": "github.com/elazarl/go-bindata-assetfs",

--- a/Godeps/LICENSES.md
+++ b/Godeps/LICENSES.md
@@ -13,7 +13,7 @@ github.com/beorn7/perks/quantile | MIT?
 github.com/blang/semver | MITname
 github.com/boltdb/bolt | MITname
 github.com/bradfitz/http2 | BSDlikeRef
-github.com/camlistore/camlistore/pkg/errorutil | Apache-2
+github.com/camlistore/go4 | Apache-2
 github.com/ClusterHQ/flocker-go | UNKNOWN
 github.com/codegangsta/negroni | MITname
 github.com/coreos/etcd | Apache-2
@@ -29,7 +29,7 @@ github.com/daviddengcn/go-colortext | BSD?
 github.com/dgrijalva/jwt-go | spdxMIT
 github.com/docker/docker | Apache-2
 github.com/docker/docker/pkg/symlink | spdxBSD3
-github.com/docker/spdystream | Apache-2
+github.com/docker/spdystream | SeeFile
 github.com/elazarl/go-bindata-assetfs | spdxBSD2
 github.com/elazarl/goproxy | BSDWarr
 github.com/emicklei/go-restful | MITname
@@ -81,6 +81,7 @@ github.com/Sirupsen/logrus | MITname
 github.com/skynetservices/skydns | MITname
 github.com/spf13/cobra | Apache-2
 github.com/spf13/pflag | spdxBSD3
+github.com/ssoroka/ttime | UNKNOWN
 github.com/stretchr/objx | MIT?
 github.com/stretchr/testify | MIT?
 github.com/syndtr/gocapability | spdxBSD2

--- a/Godeps/_workspace/src/github.com/docker/spdystream/LICENSE
+++ b/Godeps/_workspace/src/github.com/docker/spdystream/LICENSE
@@ -176,7 +176,7 @@
 
    END OF TERMS AND CONDITIONS
 
-   Copyright 2014 Docker, Inc.
+   Copyright 2014-2015 Docker, Inc.
 
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.

--- a/Godeps/_workspace/src/github.com/docker/spdystream/LICENSE.docs
+++ b/Godeps/_workspace/src/github.com/docker/spdystream/LICENSE.docs
@@ -1,0 +1,425 @@
+Attribution-ShareAlike 4.0 International
+
+=======================================================================
+
+Creative Commons Corporation ("Creative Commons") is not a law firm and
+does not provide legal services or legal advice. Distribution of
+Creative Commons public licenses does not create a lawyer-client or
+other relationship. Creative Commons makes its licenses and related
+information available on an "as-is" basis. Creative Commons gives no
+warranties regarding its licenses, any material licensed under their
+terms and conditions, or any related information. Creative Commons
+disclaims all liability for damages resulting from their use to the
+fullest extent possible.
+
+Using Creative Commons Public Licenses
+
+Creative Commons public licenses provide a standard set of terms and
+conditions that creators and other rights holders may use to share
+original works of authorship and other material subject to copyright
+and certain other rights specified in the public license below. The
+following considerations are for informational purposes only, are not
+exhaustive, and do not form part of our licenses.
+
+     Considerations for licensors: Our public licenses are
+     intended for use by those authorized to give the public
+     permission to use material in ways otherwise restricted by
+     copyright and certain other rights. Our licenses are
+     irrevocable. Licensors should read and understand the terms
+     and conditions of the license they choose before applying it.
+     Licensors should also secure all rights necessary before
+     applying our licenses so that the public can reuse the
+     material as expected. Licensors should clearly mark any
+     material not subject to the license. This includes other CC-
+     licensed material, or material used under an exception or
+     limitation to copyright. More considerations for licensors:
+	wiki.creativecommons.org/Considerations_for_licensors
+
+     Considerations for the public: By using one of our public
+     licenses, a licensor grants the public permission to use the
+     licensed material under specified terms and conditions. If
+     the licensor's permission is not necessary for any reason--for
+     example, because of any applicable exception or limitation to
+     copyright--then that use is not regulated by the license. Our
+     licenses grant only permissions under copyright and certain
+     other rights that a licensor has authority to grant. Use of
+     the licensed material may still be restricted for other
+     reasons, including because others have copyright or other
+     rights in the material. A licensor may make special requests,
+     such as asking that all changes be marked or described.
+     Although not required by our licenses, you are encouraged to
+     respect those requests where reasonable. More_considerations
+     for the public:
+	wiki.creativecommons.org/Considerations_for_licensees
+
+=======================================================================
+
+Creative Commons Attribution-ShareAlike 4.0 International Public
+License
+
+By exercising the Licensed Rights (defined below), You accept and agree
+to be bound by the terms and conditions of this Creative Commons
+Attribution-ShareAlike 4.0 International Public License ("Public
+License"). To the extent this Public License may be interpreted as a
+contract, You are granted the Licensed Rights in consideration of Your
+acceptance of these terms and conditions, and the Licensor grants You
+such rights in consideration of benefits the Licensor receives from
+making the Licensed Material available under these terms and
+conditions.
+
+
+Section 1 -- Definitions.
+
+  a. Adapted Material means material subject to Copyright and Similar
+     Rights that is derived from or based upon the Licensed Material
+     and in which the Licensed Material is translated, altered,
+     arranged, transformed, or otherwise modified in a manner requiring
+     permission under the Copyright and Similar Rights held by the
+     Licensor. For purposes of this Public License, where the Licensed
+     Material is a musical work, performance, or sound recording,
+     Adapted Material is always produced where the Licensed Material is
+     synched in timed relation with a moving image.
+
+  b. Adapter's License means the license You apply to Your Copyright
+     and Similar Rights in Your contributions to Adapted Material in
+     accordance with the terms and conditions of this Public License.
+
+  c. BY-SA Compatible License means a license listed at
+     creativecommons.org/compatiblelicenses, approved by Creative
+     Commons as essentially the equivalent of this Public License.
+
+  d. Copyright and Similar Rights means copyright and/or similar rights
+     closely related to copyright including, without limitation,
+     performance, broadcast, sound recording, and Sui Generis Database
+     Rights, without regard to how the rights are labeled or
+     categorized. For purposes of this Public License, the rights
+     specified in Section 2(b)(1)-(2) are not Copyright and Similar
+     Rights.
+
+  e. Effective Technological Measures means those measures that, in the
+     absence of proper authority, may not be circumvented under laws
+     fulfilling obligations under Article 11 of the WIPO Copyright
+     Treaty adopted on December 20, 1996, and/or similar international
+     agreements.
+
+  f. Exceptions and Limitations means fair use, fair dealing, and/or
+     any other exception or limitation to Copyright and Similar Rights
+     that applies to Your use of the Licensed Material.
+
+  g. License Elements means the license attributes listed in the name
+     of a Creative Commons Public License. The License Elements of this
+     Public License are Attribution and ShareAlike.
+
+  h. Licensed Material means the artistic or literary work, database,
+     or other material to which the Licensor applied this Public
+     License.
+
+  i. Licensed Rights means the rights granted to You subject to the
+     terms and conditions of this Public License, which are limited to
+     all Copyright and Similar Rights that apply to Your use of the
+     Licensed Material and that the Licensor has authority to license.
+
+  j. Licensor means the individual(s) or entity(ies) granting rights
+     under this Public License.
+
+  k. Share means to provide material to the public by any means or
+     process that requires permission under the Licensed Rights, such
+     as reproduction, public display, public performance, distribution,
+     dissemination, communication, or importation, and to make material
+     available to the public including in ways that members of the
+     public may access the material from a place and at a time
+     individually chosen by them.
+
+  l. Sui Generis Database Rights means rights other than copyright
+     resulting from Directive 96/9/EC of the European Parliament and of
+     the Council of 11 March 1996 on the legal protection of databases,
+     as amended and/or succeeded, as well as other essentially
+     equivalent rights anywhere in the world.
+
+  m. You means the individual or entity exercising the Licensed Rights
+     under this Public License. Your has a corresponding meaning.
+
+
+Section 2 -- Scope.
+
+  a. License grant.
+
+       1. Subject to the terms and conditions of this Public License,
+          the Licensor hereby grants You a worldwide, royalty-free,
+          non-sublicensable, non-exclusive, irrevocable license to
+          exercise the Licensed Rights in the Licensed Material to:
+
+            a. reproduce and Share the Licensed Material, in whole or
+               in part; and
+
+            b. produce, reproduce, and Share Adapted Material.
+
+       2. Exceptions and Limitations. For the avoidance of doubt, where
+          Exceptions and Limitations apply to Your use, this Public
+          License does not apply, and You do not need to comply with
+          its terms and conditions.
+
+       3. Term. The term of this Public License is specified in Section
+          6(a).
+
+       4. Media and formats; technical modifications allowed. The
+          Licensor authorizes You to exercise the Licensed Rights in
+          all media and formats whether now known or hereafter created,
+          and to make technical modifications necessary to do so. The
+          Licensor waives and/or agrees not to assert any right or
+          authority to forbid You from making technical modifications
+          necessary to exercise the Licensed Rights, including
+          technical modifications necessary to circumvent Effective
+          Technological Measures. For purposes of this Public License,
+          simply making modifications authorized by this Section 2(a)
+          (4) never produces Adapted Material.
+
+       5. Downstream recipients.
+
+            a. Offer from the Licensor -- Licensed Material. Every
+               recipient of the Licensed Material automatically
+               receives an offer from the Licensor to exercise the
+               Licensed Rights under the terms and conditions of this
+               Public License.
+
+            b. Additional offer from the Licensor -- Adapted Material.
+               Every recipient of Adapted Material from You
+               automatically receives an offer from the Licensor to
+               exercise the Licensed Rights in the Adapted Material
+               under the conditions of the Adapter's License You apply.
+
+            c. No downstream restrictions. You may not offer or impose
+               any additional or different terms or conditions on, or
+               apply any Effective Technological Measures to, the
+               Licensed Material if doing so restricts exercise of the
+               Licensed Rights by any recipient of the Licensed
+               Material.
+
+       6. No endorsement. Nothing in this Public License constitutes or
+          may be construed as permission to assert or imply that You
+          are, or that Your use of the Licensed Material is, connected
+          with, or sponsored, endorsed, or granted official status by,
+          the Licensor or others designated to receive attribution as
+          provided in Section 3(a)(1)(A)(i).
+
+  b. Other rights.
+
+       1. Moral rights, such as the right of integrity, are not
+          licensed under this Public License, nor are publicity,
+          privacy, and/or other similar personality rights; however, to
+          the extent possible, the Licensor waives and/or agrees not to
+          assert any such rights held by the Licensor to the limited
+          extent necessary to allow You to exercise the Licensed
+          Rights, but not otherwise.
+
+       2. Patent and trademark rights are not licensed under this
+          Public License.
+
+       3. To the extent possible, the Licensor waives any right to
+          collect royalties from You for the exercise of the Licensed
+          Rights, whether directly or through a collecting society
+          under any voluntary or waivable statutory or compulsory
+          licensing scheme. In all other cases the Licensor expressly
+          reserves any right to collect such royalties.
+
+
+Section 3 -- License Conditions.
+
+Your exercise of the Licensed Rights is expressly made subject to the
+following conditions.
+
+  a. Attribution.
+
+       1. If You Share the Licensed Material (including in modified
+          form), You must:
+
+            a. retain the following if it is supplied by the Licensor
+               with the Licensed Material:
+
+                 i. identification of the creator(s) of the Licensed
+                    Material and any others designated to receive
+                    attribution, in any reasonable manner requested by
+                    the Licensor (including by pseudonym if
+                    designated);
+
+                ii. a copyright notice;
+
+               iii. a notice that refers to this Public License;
+
+                iv. a notice that refers to the disclaimer of
+                    warranties;
+
+                 v. a URI or hyperlink to the Licensed Material to the
+                    extent reasonably practicable;
+
+            b. indicate if You modified the Licensed Material and
+               retain an indication of any previous modifications; and
+
+            c. indicate the Licensed Material is licensed under this
+               Public License, and include the text of, or the URI or
+               hyperlink to, this Public License.
+
+       2. You may satisfy the conditions in Section 3(a)(1) in any
+          reasonable manner based on the medium, means, and context in
+          which You Share the Licensed Material. For example, it may be
+          reasonable to satisfy the conditions by providing a URI or
+          hyperlink to a resource that includes the required
+          information.
+
+       3. If requested by the Licensor, You must remove any of the
+          information required by Section 3(a)(1)(A) to the extent
+          reasonably practicable.
+
+  b. ShareAlike.
+
+     In addition to the conditions in Section 3(a), if You Share
+     Adapted Material You produce, the following conditions also apply.
+
+       1. The Adapter's License You apply must be a Creative Commons
+          license with the same License Elements, this version or
+          later, or a BY-SA Compatible License.
+
+       2. You must include the text of, or the URI or hyperlink to, the
+          Adapter's License You apply. You may satisfy this condition
+          in any reasonable manner based on the medium, means, and
+          context in which You Share Adapted Material.
+
+       3. You may not offer or impose any additional or different terms
+          or conditions on, or apply any Effective Technological
+          Measures to, Adapted Material that restrict exercise of the
+          rights granted under the Adapter's License You apply.
+
+
+Section 4 -- Sui Generis Database Rights.
+
+Where the Licensed Rights include Sui Generis Database Rights that
+apply to Your use of the Licensed Material:
+
+  a. for the avoidance of doubt, Section 2(a)(1) grants You the right
+     to extract, reuse, reproduce, and Share all or a substantial
+     portion of the contents of the database;
+
+  b. if You include all or a substantial portion of the database
+     contents in a database in which You have Sui Generis Database
+     Rights, then the database in which You have Sui Generis Database
+     Rights (but not its individual contents) is Adapted Material,
+
+     including for purposes of Section 3(b); and
+  c. You must comply with the conditions in Section 3(a) if You Share
+     all or a substantial portion of the contents of the database.
+
+For the avoidance of doubt, this Section 4 supplements and does not
+replace Your obligations under this Public License where the Licensed
+Rights include other Copyright and Similar Rights.
+
+
+Section 5 -- Disclaimer of Warranties and Limitation of Liability.
+
+  a. UNLESS OTHERWISE SEPARATELY UNDERTAKEN BY THE LICENSOR, TO THE
+     EXTENT POSSIBLE, THE LICENSOR OFFERS THE LICENSED MATERIAL AS-IS
+     AND AS-AVAILABLE, AND MAKES NO REPRESENTATIONS OR WARRANTIES OF
+     ANY KIND CONCERNING THE LICENSED MATERIAL, WHETHER EXPRESS,
+     IMPLIED, STATUTORY, OR OTHER. THIS INCLUDES, WITHOUT LIMITATION,
+     WARRANTIES OF TITLE, MERCHANTABILITY, FITNESS FOR A PARTICULAR
+     PURPOSE, NON-INFRINGEMENT, ABSENCE OF LATENT OR OTHER DEFECTS,
+     ACCURACY, OR THE PRESENCE OR ABSENCE OF ERRORS, WHETHER OR NOT
+     KNOWN OR DISCOVERABLE. WHERE DISCLAIMERS OF WARRANTIES ARE NOT
+     ALLOWED IN FULL OR IN PART, THIS DISCLAIMER MAY NOT APPLY TO YOU.
+
+  b. TO THE EXTENT POSSIBLE, IN NO EVENT WILL THE LICENSOR BE LIABLE
+     TO YOU ON ANY LEGAL THEORY (INCLUDING, WITHOUT LIMITATION,
+     NEGLIGENCE) OR OTHERWISE FOR ANY DIRECT, SPECIAL, INDIRECT,
+     INCIDENTAL, CONSEQUENTIAL, PUNITIVE, EXEMPLARY, OR OTHER LOSSES,
+     COSTS, EXPENSES, OR DAMAGES ARISING OUT OF THIS PUBLIC LICENSE OR
+     USE OF THE LICENSED MATERIAL, EVEN IF THE LICENSOR HAS BEEN
+     ADVISED OF THE POSSIBILITY OF SUCH LOSSES, COSTS, EXPENSES, OR
+     DAMAGES. WHERE A LIMITATION OF LIABILITY IS NOT ALLOWED IN FULL OR
+     IN PART, THIS LIMITATION MAY NOT APPLY TO YOU.
+
+  c. The disclaimer of warranties and limitation of liability provided
+     above shall be interpreted in a manner that, to the extent
+     possible, most closely approximates an absolute disclaimer and
+     waiver of all liability.
+
+
+Section 6 -- Term and Termination.
+
+  a. This Public License applies for the term of the Copyright and
+     Similar Rights licensed here. However, if You fail to comply with
+     this Public License, then Your rights under this Public License
+     terminate automatically.
+
+  b. Where Your right to use the Licensed Material has terminated under
+     Section 6(a), it reinstates:
+
+       1. automatically as of the date the violation is cured, provided
+          it is cured within 30 days of Your discovery of the
+          violation; or
+
+       2. upon express reinstatement by the Licensor.
+
+     For the avoidance of doubt, this Section 6(b) does not affect any
+     right the Licensor may have to seek remedies for Your violations
+     of this Public License.
+
+  c. For the avoidance of doubt, the Licensor may also offer the
+     Licensed Material under separate terms or conditions or stop
+     distributing the Licensed Material at any time; however, doing so
+     will not terminate this Public License.
+
+  d. Sections 1, 5, 6, 7, and 8 survive termination of this Public
+     License.
+
+
+Section 7 -- Other Terms and Conditions.
+
+  a. The Licensor shall not be bound by any additional or different
+     terms or conditions communicated by You unless expressly agreed.
+
+  b. Any arrangements, understandings, or agreements regarding the
+     Licensed Material not stated herein are separate from and
+     independent of the terms and conditions of this Public License.
+
+
+Section 8 -- Interpretation.
+
+  a. For the avoidance of doubt, this Public License does not, and
+     shall not be interpreted to, reduce, limit, restrict, or impose
+     conditions on any use of the Licensed Material that could lawfully
+     be made without permission under this Public License.
+
+  b. To the extent possible, if any provision of this Public License is
+     deemed unenforceable, it shall be automatically reformed to the
+     minimum extent necessary to make it enforceable. If the provision
+     cannot be reformed, it shall be severed from this Public License
+     without affecting the enforceability of the remaining terms and
+     conditions.
+
+  c. No term or condition of this Public License will be waived and no
+     failure to comply consented to unless expressly agreed to by the
+     Licensor.
+
+  d. Nothing in this Public License constitutes or may be interpreted
+     as a limitation upon, or waiver of, any privileges and immunities
+     that apply to the Licensor or You, including from the legal
+     processes of any jurisdiction or authority.
+
+
+=======================================================================
+
+Creative Commons is not a party to its public licenses.
+Notwithstanding, Creative Commons may elect to apply one of its public
+licenses to material it publishes and in those instances will be
+considered the "Licensor." Except for the limited purpose of indicating
+that material is shared under a Creative Commons public license or as
+otherwise permitted by the Creative Commons policies published at
+creativecommons.org/policies, Creative Commons does not authorize the
+use of the trademark "Creative Commons" or any other trademark or logo
+of Creative Commons without its prior written consent including,
+without limitation, in connection with any unauthorized modifications
+to any of its public licenses or any other arrangements,
+understandings, or agreements concerning use of licensed material. For
+the avoidance of doubt, this paragraph does not form part of the public
+licenses.
+
+Creative Commons may be contacted at creativecommons.org.

--- a/Godeps/_workspace/src/github.com/docker/spdystream/MAINTAINERS
+++ b/Godeps/_workspace/src/github.com/docker/spdystream/MAINTAINERS
@@ -1,1 +1,28 @@
-Derek McGowan <derek@docker.com> (@dmcg)
+# Spdystream maintainers file
+#
+# This file describes who runs the docker/spdystream project and how.
+# This is a living document - if you see something out of date or missing, speak up!
+#
+# It is structured to be consumable by both humans and programs.
+# To extract its contents programmatically, use any TOML-compliant parser.
+#
+# This file is compiled into the MAINTAINERS file in docker/opensource.
+#
+[Org]
+	[Org."Core maintainers"]
+		people = [
+			"dmcgowan",
+		]
+
+[people]
+
+# A reference list of all people associated with the project.
+# All other sections should refer to people by their canonical key
+# in the people section.
+
+	# ADD YOURSELF HERE IN ALPHABETICAL ORDER
+
+	[people.dmcgowan]
+	Name = "Derek McGowan"
+	Email = "derek@docker.com"
+	GitHub = "dmcgowan"

--- a/Godeps/_workspace/src/github.com/docker/spdystream/README.md
+++ b/Godeps/_workspace/src/github.com/docker/spdystream/README.md
@@ -74,5 +74,4 @@ func main() {
 
 ## Copyright and license
 
-Code and documentation copyright 2013-2014 Docker, inc. Code released under the Apache 2.0 license.
-Docs released under Creative commons.
+Copyright © 2014-2015 Docker, Inc. All rights reserved, except as follows. Code is released under the Apache 2.0 license. The README.md file, and files in the "docs" folder are licensed under the Creative Commons Attribution 4.0 International License under the terms and conditions set forth in the file "LICENSE.docs". You may obtain a duplicate copy of the same license, titled CC-BY-SA-4.0, at http://creativecommons.org/licenses/by/4.0/.

--- a/Godeps/_workspace/src/github.com/docker/spdystream/connection.go
+++ b/Godeps/_workspace/src/github.com/docker/spdystream/connection.go
@@ -33,24 +33,28 @@ type idleAwareFramer struct {
 	conn           *Connection
 	writeLock      sync.Mutex
 	resetChan      chan struct{}
+	setTimeoutLock sync.Mutex
 	setTimeoutChan chan time.Duration
 	timeout        time.Duration
 }
 
 func newIdleAwareFramer(framer *spdy.Framer) *idleAwareFramer {
 	iaf := &idleAwareFramer{
-		f:              framer,
-		resetChan:      make(chan struct{}, 2),
-		setTimeoutChan: make(chan time.Duration),
+		f:         framer,
+		resetChan: make(chan struct{}, 2),
+		// setTimeoutChan needs to be buffered to avoid deadlocks when calling setIdleTimeout at about
+		// the same time the connection is being closed
+		setTimeoutChan: make(chan time.Duration, 1),
 	}
 	return iaf
 }
 
 func (i *idleAwareFramer) monitor() {
 	var (
-		timer     *time.Timer
-		expired   <-chan time.Time
-		resetChan = i.resetChan
+		timer          *time.Timer
+		expired        <-chan time.Time
+		resetChan      = i.resetChan
+		setTimeoutChan = i.setTimeoutChan
 	)
 Loop:
 	for {
@@ -103,10 +107,20 @@ Loop:
 				}
 			}()
 
+			go func() {
+				for _ = range setTimeoutChan {
+				}
+			}()
+
 			i.writeLock.Lock()
 			close(resetChan)
 			i.resetChan = nil
 			i.writeLock.Unlock()
+
+			i.setTimeoutLock.Lock()
+			close(i.setTimeoutChan)
+			i.setTimeoutChan = nil
+			i.setTimeoutLock.Unlock()
 
 			break Loop
 		}
@@ -146,6 +160,17 @@ func (i *idleAwareFramer) ReadFrame() (spdy.Frame, error) {
 	i.resetChan <- struct{}{}
 
 	return frame, nil
+}
+
+func (i *idleAwareFramer) setIdleTimeout(timeout time.Duration) {
+	i.setTimeoutLock.Lock()
+	defer i.setTimeoutLock.Unlock()
+
+	if i.setTimeoutChan == nil {
+		return
+	}
+
+	i.setTimeoutChan <- timeout
 }
 
 type Connection struct {
@@ -794,7 +819,7 @@ func (s *Connection) SetCloseTimeout(timeout time.Duration) {
 // SetIdleTimeout sets the amount of time the connection may sit idle before
 // it is forcefully terminated.
 func (s *Connection) SetIdleTimeout(timeout time.Duration) {
-	s.framer.setTimeoutChan <- timeout
+	s.framer.setIdleTimeout(timeout)
 }
 
 func (s *Connection) sendHeaders(headers http.Header, stream *Stream, fin bool) error {

--- a/Godeps/_workspace/src/github.com/docker/spdystream/stream.go
+++ b/Godeps/_workspace/src/github.com/docker/spdystream/stream.go
@@ -322,6 +322,5 @@ func (s *Stream) closeRemoteChannels() {
 		close(s.closeChan)
 		s.dataLock.Lock()
 		defer s.dataLock.Unlock()
-		close(s.dataChan)
 	}
 }

--- a/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt
+++ b/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt
@@ -1,0 +1,9 @@
+Copyright (c) 2012 - 2013 Mat Ryer and Tyler Bunnell
+
+Please consider promoting this project if you find it useful.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

--- a/hack/update-godep-licenses.sh
+++ b/hack/update-godep-licenses.sh
@@ -41,4 +41,5 @@ exec docker run --rm -i -v "${KUBE_ROOT}:/repo" mesosphere/godep-licenses:latest
   -e github.com/spf13/cobra:Apache-2 \
   -e github.com/stretchr/objx:MIT? \
   -e github.com/stretchr/testify:MIT? \
+  -e github.com/docker/spdystream:SeeFile \
   -o md > Godeps/LICENSES.md

